robot_3d/src/joint_states_subscriber_3d.cpp

```diff
@@ -0,0 +1,233 @@
+
+#include <boost/filesystem.hpp>
+
+#include "yaml-cpp/yaml.h"
+
+#include <string>
+#include <sstream>
+#include <vector>
+
+class JointMeasurementData
+{
+public:
+  JointMeasurementData(ros::NodeHandle* n);
+
+private:
+  void jointStatesCallback(const sensor_msgs::JointState::ConstPtr& msg);
+  // Method to calculate the end effector position using kdl and broadcast the transform.
+  void kdlCalc(double joint_positions[]);
+  // Method to output the pose of the end effector position to yaml file.
+  void outputYaml(tf::Transform kdl_transform, double joint_positions[]);
+  std::vector<double> getJointOffsets(double joint_positions[]);
+
+  ros::Subscriber sub_;
+
+  // String to store the filepath for yaml file.
+  std::string output_filepath_;
+  // String to store data to be sent to yaml file.
+  std::string output_data_ = "Data: \n";
+  // The number of data points accepted by node until it terminates.
+  int num_data_point_;
+  // Count of the data points.
+  int data_point_count_;
+};
+
+JointMeasurementData::JointMeasurementData(ros::NodeHandle* n)
+{
+  // Sets the count to zero.
+  data_point_count_ = 0;
+
+  sub_ = n->subscribe("joint_states", 1000, &JointMeasurementData::jointStatesCallback, this);
+
+  // Gets the maximum number of data points from launch file.
+  n->getParam("data_point_count", num_data_point_);
+
+  // Gets the currrent date and time to include in the output filename.
+  char date_holder[21];
+  time_t now;
+  now = time(0);
+  if (now != -1)
+  {
+    strftime(date_holder, 20, "%d_%m_%Y_%T", gmtime(&now));
+  }
+
+  // Gets the robot_3d package filepath.
+  output_filepath_ = ros::package::getPath("robot_3d");
+
+  // Checks if Output_yaml folder exists in package and creates folder if it does not already exist.
+  std::string check = output_filepath_.append("/Output_yaml");
+  if (!(boost::filesystem::exists(check)))
+  {
+    boost::filesystem::create_directory(check);
+  }
+  // Amends filepath for storing yaml file.
+  output_filepath_ = check;
+  output_filepath_ = output_filepath_ + "/" + std::string(date_holder);
+  output_filepath_.append("_output.yaml");
+}
+
+void JointMeasurementData::jointStatesCallback(const sensor_msgs::JointState::ConstPtr& msg)
+{
+  ROS_INFO("I heard: [%F]", msg->position[0]);
+
+  // Increments the data point count.
+  data_point_count_++;
+
+  // Array to store the six joint positions from the publisher node.
+  double position[6];
+  position[0] = msg->position[0];
+  position[1] = msg->position[1];
+  position[2] = msg->position[2];
+  position[3] = msg->position[3];
+  position[4] = msg->position[4];
+  position[5] = msg->position[5];
+
+  // Passing joint positons to method to calculate end effector position.
+  kdlCalc(position);
+
+  // Terminates node if the maximum number of data points is achieved.
+  if (data_point_count_ == num_data_point_)
+  {
+    ros::shutdown();
+  }
+}
+
+void JointMeasurementData::kdlCalc(double joint_positions[])
```

Comment:
FYI: You will have to rebase this branch so that you get the changes I said to make on the KDL branch.

Reply:
...does it count as cheating if I already included at least some of the changes from the kdl branch since I would have to amend  a couple other things in this branch?